Src/FileIndexThread.pas

```diff
@@ -0,0 +1,119 @@
+unit FileIndexThread;
+
+interface
+
+uses
+  System.Classes,
+  System.SyncObjs,
+  System.SysUtils,
+  Services.IRepo;
+
+type
+  TFileIndexThread = Class(TThread)
+    strict private
+      FEvent: TEvent;
+      FRepo: IRepo;
+
+      procedure IndexFiles;
+    protected
+      procedure Execute; override;
+    public
+      constructor Create(const aRepo: IRepo); overload;
+      destructor Destroy; override;
+  End;
+
+
+implementation
+
+uses
+  ToolsAPI,
+  System.IOUtils,
+  Spring.Collections;
+
+{ TFileIndexThread }
+
+constructor TFileIndexThread.Create(const aRepo: IRepo);
+begin
+  Inherited Create(true);
+  FRepo := aRepo;
+
+  Suspended := false;
+end;
+
+destructor TFileIndexThread.Destroy;
+begin
+  FreeAndNil(FEvent);
+  inherited;
+end;
+
+procedure TFileIndexThread.Execute;
+begin
+  inherited;
+
+  FEvent := TEvent.Create(nil, True, False, '');
+  while True do begin
+
+    IndexFiles;
+
+
+    if Terminated then Break;
+    try FEvent.WaitFor(3000); except end; //Wait for 3 seconds
+    if Terminated then Break;
+  end;
+end;
+
+
+procedure TFileIndexThread.IndexFiles;
+begin
+  var lProject := GetActiveProject;
+  if not Assigned(lProject) then //Should only try and find files, if a project is currently active. Exit early otherwise
+    Exit;
+
+  var lProjectPath := ExtractFilePath(lProject.FileName);
+
+  var lFiles := TCollections.CreateList<string>;
+
+  var lTmpFiles := TStringList.Create;
+  lProject.GetCompleteFileList(lTmpFiles);
+
+  for var lFile in lTmpFiles do
+  begin
+    if ExtractFileExt(lFile) = '.pas' then
+    begin
+      var lRelativePath := ExtractRelativePath(lProjectPath, lFile);
+      if not lFiles.Contains(lRelativePath) then
+        lFiles.Add(lRelativePath);
```

Comment:
A `set` would probably be a more optimal data structure for `lFiles` since it inherently, each element can only occur once, so you would be able to save that `if not contains` check. If there are a large number of files, this would probably not be an insignificant computational saving?

Reply:
True using a `TDictionary` would mean that the lookup is O(1) instead of a O(n).
So could definitely save some time.

Reply:
Turns out there are `sets` When using Spring4D so i changed it to use that 😄